### PR TITLE
Added repository manipulation functions

### PR DESCRIPTION
**Pull request by rkesters**
I added the ability to manipulate a git repo

- checkout branch
- create local and remote branches
- delete local and remote branches
- create tags
- push

### CODE REVIEW

**Comment by rkesters** [comment]
@septs can you review and publish to npm?

File: src/git.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
-import { execFileSync, ExecFileSyncOptionsWithStringEncoding } from "child_process";
-import { basename } from "path";
+import {execFileSync, ExecFileSyncOptionsWithStringEncoding} from "child_process";
+import {basename} from "path";
```
Comment (septs) [inline src/git.ts]: Please keep tslint formatter code style

File: src/git.ts [hunk 2]
```diff
@@ -63,19 +63,20 @@ export class Git {
...
     }
 
-    public tag(markDirty = false, firstParent = false) {
+    public tag(markDirty = false, firstParent = false, match?: string) {
+        const matchArgs = match ? ['--match', match] : [];
         return this.describe(
             "--always", "--tag", "--abbrev=0",
             markDirty && "--dirty",
-            firstParent && "--first-parent"
+            firstParent && "--first-parent", ...matchArgs
```
Comment (septs) [inline src/git.ts]: Here I need to refactor to options pattern.

File: src/git.ts [hunk 3]
```diff
@@ -124,4 +125,62 @@ export class Git {
         const url = this.remoteURL(remoteName);
         return basename(url, ".git");
     }
+
```
Comment (septs) [inline src/git.ts]: I don't want to manipulate the git repository.

But consider providing a separate file

```typescript
import { GitAgent } from "@nice-labs/git-rev";
const agent = GitAgent(process.cwd());
agent.commit(...);
agent.createBranch(...);
```

File: tests/setup.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env sh
```
Comment (septs) [inline tests/setup.sh]: We should consider **Windows** platform.

File: src/git.ts (hunk 2 above)
Comment (septs) [reply src/git.ts]: I think I should probably add preprocessing like npmjs.com/classnames

File: src/git.ts (hunk 3 above)
Comment (septs) [reply src/git.ts]:
```typescript
// src/git-agent.ts
export class GitAgent extends Git {
    // ... your manipulate the git repository logic ...
}

// before: update index.ts expose GitAgent
```

File: tests/setup.sh (hunk 4 above)
Comment (septs) [reply tests/setup.sh]: https://git-scm.com/docs/git-bundle

File: src/git.ts [hunk 5]
```diff
@@ -124,4 +125,62 @@ export class Git {
...
+    public deleteBranch(name: string) {
+        return this.git( 'branch' , '-d'  ,   name)
+    }
+
+    public createTag(tag: string) {
+        this.git('tag', tag);
+    }
+
+    public push(origin: string = 'origin', noVerify: boolean = true,
```
Comment (septs) [inline src/git.ts]:
```suggestion
    public push(origin: string = 'origin', noVerify: boolean = true,
    public push(options: PushOptions)
```

File: src/git.ts [hunk 6]
```diff
@@ -63,19 +63,20 @@ export class Git {
         return this.git("config", name, value);
     }
 
-    public tag(markDirty = false, firstParent = false) {
+    public tag(markDirty = false, firstParent = false, match?: string) {
```
Comment (septs) [inline src/git.ts]:
```suggestion
    public tag(options: TagOptions) {
```

**Comment by septs** [comment]
@rkesters 
https://github.com/NiceLabs/git-rev/commit/e3c6e53d016a70d69ec017f8c2dcbfb64fc2cd37

**Comment by septs** [comment]
release 3.3.0
npmjs.com/@nice-labs/git-rev

**Pull request closed without merging**